File: assignments/week-11/1036-rotting-oranges/solution.java
class Pair<R,C,M>{
    R rowIdx;
    C colIdx;
    M minutesCount;

    Pair(R rowIdx, C colIdx, M minutesCount){
        this.rowIdx = rowIdx;
        this.colIdx = colIdx;
        this.minutesCount = minutesCount;
    }

};

class Solution {
    public int orangesRotting(int[][] grid) {
        int mRows = grid.length;
        int nCols = grid[0].length;
        
        int minMinutesCount = 0;
        int totalFreshOranges = 0;
        int totalRottenOranges = 0;

        Deque<Pair<Integer, Integer, Integer>> queue = new ArrayDeque<>();
        int[][] rottenOrangesMap = new int[mRows][nCols];

        // right, bottom, left, and top;
        int[] dx = new int[] {1, 0, -1, 0};
        int[] dy = new int[] {0, 1, 0, -1};

        for(int[] currRow : rottenOrangesMap){
            Arrays.fill(currRow, 2);
        }

        for(int rowIdx = 0; rowIdx < mRows; rowIdx++){
            for(int colIdx = 0; colIdx < nCols; colIdx++){
                if(grid[rowIdx][colIdx] == 1){
                    totalFreshOranges += 1;
                    rottenOrangesMap[rowIdx][colIdx] = 1;
                }
                else if(grid[rowIdx][colIdx] == 2){
                    queue.addLast(new Pair<>(rowIdx, colIdx, 0));
                }
            }
        }

        while(queue.size() > 0){
            Pair<Integer, Integer, Integer> firstEl = queue.removeFirst();

            int rowIdx = firstEl.rowIdx;
            int colIdx = firstEl.colIdx;
            int minutesCount = firstEl.minutesCount;

            minMinutesCount = minutesCount;

            for(int idx = 0; idx < 4; idx++){
                int newRowIdx = rowIdx + dx[idx];
                int newColIdx = colIdx + dy[idx];

                if(newRowIdx >= 0 && newRowIdx < mRows && newColIdx >= 0 && newColIdx < nCols && grid[newRowIdx][newColIdx] == 1 && rottenOrangesMap[newRowIdx][newColIdx] == 1){
                    rottenOrangesMap[newRowIdx][newColIdx] = 2;
                    queue.addLast(new Pair<>(newRowIdx, newColIdx, minutesCount + 1));
                    totalRottenOranges += 1;
                }
            }
        }

        // System.out.println(totalFreshOranges + "" + totalRottenOranges);
        // System.out.println(minMinutesCount);
        /**
        if(totalFreshOranges == 0){
            return 0;
        }
        else if(totalFreshOranges == totalRottenOranges){
            return minMinutesCount;
        }

        return -1;
        */

        if(totalFreshOranges == 0){
            return 0;
        }
        else if(totalFreshOranges > totalRottenOranges){
            return -1;
        }

        return minMinutesCount;
    }
}

File: assignments/week-11/1036-rotting-oranges/solution.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m_rows = len(grid)
        n_cols = len(grid[0])

        queue = []

        rotten_oranges = [[-1 for _ in range(n_cols)] for _ in range(m_rows)]
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        total_fresh_oranges = 0

        for row_idx in range(m_rows):
            for col_idx in range(n_cols):
                if grid[row_idx][col_idx] == 2:
                    queue.append((row_idx, col_idx, 0))
                    rotten_oranges[row_idx][col_idx] = 2
                elif grid[row_idx][col_idx] == 0:
                    rotten_oranges[row_idx][col_idx] = 2
                else:
                    total_fresh_oranges += 1
        
        min_mins_count = 0
        # print(total_fresh_oranges, rotten_oranges)

        while len(queue) > 0:
            # print(queue)
            row_idx, col_idx, curr_time = queue.pop(0)
            min_mins_count = max(min_mins_count, curr_time)
            
            for direction in directions:
                new_row_idx = row_idx + direction[0]
                new_col_idx = col_idx + direction[1]

                if new_row_idx >= 0 and new_row_idx < m_rows and new_col_idx >= 0 and new_col_idx < n_cols and grid[new_row_idx][new_col_idx] == 1 and rotten_oranges[new_row_idx][new_col_idx] == -1:
                    rotten_oranges[new_row_idx][new_col_idx] = 2
                    queue.append((new_row_idx, new_col_idx, curr_time + 1))
                    total_fresh_oranges -= 1
            
        if total_fresh_oranges == 0:
            return min_mins_count 
        else:
            return -1